src/librustc/traits/project.rs

```diff
@@ -1129,10 +1135,18 @@ fn assemble_candidates_from_impls<'cx, 'gcx, 'tcx>(
                     "Cannot project an associated type from `{:?}`",
                     vtable);
             }
-        }
+        };
 
-        Ok(())
-    })
+        if eligible {
+            if candidate_set.push_candidate(ProjectionTyCandidate::Select(vtable)) {
+                Ok(())
+            } else {
+                Err(())
+            }
+        } else {
+            Err(())
+        }
+    });
```

Comment:
I like the change to push errors into the candidate set. This is less confusing this way.

Reply:
I'm not sure if I understand. I already handle SelectionError above, and these Ok and Err is the return value to `commit_if_ok`.